src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java

```diff
@@ -0,0 +1,367 @@
+     * @param exerciseId the id of the exercise
+     * @param participantId the id of the participant (user or team, determined by the exercise)
+     * @param resultLastModified the last modified date of the result that triggered the update
+     * @param resultIdsToIgnore a list of result ids that should be ignored (e.g., as they are about to be deleted)
+     */
+    private void scheduleTask(Long exerciseId, Long participantId, Instant resultLastModified, Long... resultIdsToIgnore) {
+        var participantScoreId = new ParticipantScoreId(exerciseId, participantId);
+        if (scheduledTasks.containsKey(participantScoreId.hashCode())) {
+            // We already have a scheduled task for this result
+            return;
+        }
+
+        var schedulingTime = ZonedDateTime.now().plus(500, ChronoUnit.MILLIS);
+        var future = scheduler.schedule(() -> this.executeTask(exerciseId, participantId, resultLastModified, resultIdsToIgnore), schedulingTime.toInstant());
+        scheduledTasks.put(participantScoreId.hashCode(), future);
+        logger.debug("Schedule task for participation {} at {}.", exerciseId, schedulingTime);
+    }
+
+    /**
+     * Execute the task to update the participant score for the given combination of exercise and participant.
+     * @param exerciseId the id of the exercise
+     * @param participantId the id of the participant (user or team, determined by the exercise)
+     * @param resultLastModified the last modified date of the result that triggered the update
+     * @param resultIdsToIgnore a list of result ids that should be ignored when updating the participant score
+     */
+    private void executeTask(Long exerciseId, Long participantId, Instant resultLastModified, Long... resultIdsToIgnore) {
+        logger.debug("Processing exercise {} and participant {} to update participant scores.", exerciseId, participantId);
+        try {
+            SecurityUtils.setAuthorizationObject();
+
+            var exercise = exerciseRepository.findById(exerciseId).orElse(null);
+
+            if (exercise == null) {
+                logger.debug("Exercise {} no longer exists, deleting participant score.", exerciseId);
+                participantScoreRepository.deleteAllByExerciseIdTransactional(exerciseId);
+                return;
+            }
+
+            Participant participant;
+            Optional<ParticipantScore> participantScore;
+            if (exercise.isTeamMode()) {
+                participant = teamRepository.findByIdElseThrow(participantId);
+                participantScore = teamScoreRepository.findByExercise_IdAndTeam_Id(exerciseId, participantId).map(score -> score);
+            }
+            else {
+                participant = userRepository.findByIdElseThrow(participantId);
+                participantScore = studentScoreRepository.findByExercise_IdAndUser_Id(exerciseId, participantId).map(score -> score);
+            }
+
+            if (participantScore.isPresent()) {
```

Comment:
💡 This method is rather long. Maybe you could extract some functionality into smaller methods? For example this part into a method like `checkIfUpdateIsNeeded` or something.

Reply:
It might be long, but it is not really complex (only one nested if-statement). If other reviewers agree, I'd like to keep it as-is.

Reply:
+1 for splitting up `executeTask` and extracting parts into smaller helper methods from my side.

Reply:
Although smaller methods are in general better, I don't think it adds much value here